provisioning/service/src/HttpDeviceRegistrationClient.cs

```diff
@@ -0,0 +1,660 @@
+                }
+            }
+        }
+
+        void EnsureInstanceNotClosed()
+        {
+            if (this.httpClientHelper == null)
+            {
+                throw new ObjectDisposedException("DeviceRegistrationClient", ApiResources.DeviceRegistrationClientAlreadyClosed);
+            }
+        }
+
+        public override Task<Enrollment> AddEnrollmentAsync(Enrollment enrollment)
+        {
+            return this.AddEnrollmentAsync(enrollment, CancellationToken.None);
+        }
+
+        public override Task<Enrollment> AddEnrollmentAsync(Enrollment enrollment, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+            ValidateRegistrationAndDeviceId(enrollment);
+
+            var errorMappingOverrides = new Dictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>
+            {
+                {
+                    HttpStatusCode.Conflict,
+                    responseMessage => Task.FromResult((Exception) new EnrollmentAlreadyExistsException(enrollment.RegistrationId))
+                }
+            };
+
+            return this.httpClientHelper.PutAsync(GetEnrollmentUri(enrollment.RegistrationId), enrollment, PutOperationType.CreateEntity, errorMappingOverrides, cancellationToken);
+        }
+
+        public override Task<BulkOperationResult> AddEnrollmentsAsync(IEnumerable<Enrollment> enrollments)
+        {
+            return this.AddEnrollmentsAsync(enrollments, CancellationToken.None);
+        }
+
+        public override Task<BulkOperationResult> AddEnrollmentsAsync(IEnumerable<Enrollment> enrollments, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+
+            foreach (Enrollment enrollment in enrollments)
+            {
+                ValidateRegistrationAndDeviceId(enrollment);
+            }
+
+            var operation = new BulkOperation()
+            {
+                Mode = BulkOperationMode.Create,
+                Enrollments = enrollments
+            };
+
+            return this.httpClientHelper.PostAsync<BulkOperation, BulkOperationResult>(GetBulkEnrollmentUri(), operation,
+                new Dictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>(), null, cancellationToken);
+        }
+
+        public override Task<Enrollment> GetEnrollmentAsync(string registrationId)
+        {
+            return this.GetEnrollmentAsync(registrationId, CancellationToken.None);
+        }
+
+        public override Task<Enrollment> GetEnrollmentAsync(string registrationId, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+            var errorMappingOverrides = new Dictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>
+            {
+                {
+                    HttpStatusCode.NotFound,
+                    responseMessage => Task.FromResult((Exception) new EnrollmentNotFoundException(registrationId))
```

Comment:
Perf: Allocating this exception beforehand is guaranteeing that we always consume the memory. Instead, we could allocate only when we need to throw that exception.

Reply:
#176 to track

---
In reply to: [141722179](https://github.com/Azure/azure-iot-sdk-csharp/pull/174#discussion_r141722179) [](ancestors = 141722179)